src/process.rs

```diff
@@ -330,27 +330,65 @@ fn read_process_pacs_from_iter<I>(
     iter: I,
     process_ids: &HashSet<Rc<str>>,
     commodities: &HashMap<Rc<str>, Rc<Commodity>>,
+    flows: &HashMap<Rc<str>, Vec<ProcessFlow>>,
 ) -> Result<HashMap<Rc<str>, Vec<Rc<Commodity>>>>
 where
     I: Iterator<Item = ProcessPAC>,
 {
     // Keep track of previous PACs so we can check for duplicates
-    let mut pacs = HashSet::new();
-
-    iter.map(|pac| {
-        let process_id = process_ids.get_id(&pac.process_id)?;
-        let commodity = commodities.get(pac.commodity_id.as_str());
-
-        match commodity {
-            None => bail!("{} is not a valid commodity ID", &pac.commodity_id),
-            Some(commodity) => {
-                ensure!(pacs.insert(pac), "Duplicate PACs found");
-
-                Ok((process_id, Rc::clone(commodity)))
+    let mut existing_pacs = HashSet::new();
+
+    // Build hashmap of process ID to PAC commodities
+    let pacs = iter
+        .map(|pac| {
+            let process_id = process_ids.get_id(&pac.process_id)?;
+            let commodity = commodities.get(pac.commodity_id.as_str());
+
+            // Check that commodity is valid and PAC is not a duplicate
+            match commodity {
+                None => bail!("{} is not a valid commodity ID", &pac.commodity_id),
+                Some(commodity) => {
+                    ensure!(existing_pacs.insert(pac), "Duplicate PACs found");
+                    Ok((process_id, Rc::clone(commodity)))
+                }
             }
+        })
+        .process_results(|iter| iter.into_group_map())?;
+
+    // Check that PACs for each process are either all inputs or all outputs
+    for (process_id, pacs) in pacs.iter() {
+        let flows = flows.get(process_id).unwrap();
+
+        let mut flow_sign: Option<bool> = None; // False for inputs, true for outputs
+        for pac in pacs.iter() {
+            // Find the flow associated with the PAC
+            let flow = flows
+                .iter()
+                .find(|item| *item.commodity_id.as_str() == *pac.id)
+                .with_context(|| {
+                    format!(
+                        "PAC {} for process {} must have an associated flow",
+                        pac.id, process_id
+                    )
+                })?;
+
+            // Check that flow sign is consistent
+            let current_flow_sign = flow.flow > 0.0;
+            match flow_sign {
+                None => {
+                    flow_sign = Some(current_flow_sign);
+                }
+                Some(flow_sign) => {
+                    ensure!(
+                        current_flow_sign == flow_sign,
+                        "PACs for process {} are a mix of inputs and outputs",
+                        process_id
+                    );
+                }
+            };
```

Comment:
Again, totally fine, but you could do an `if let` here instead:

```suggestion
            if let Some(flow_sign) = flow_sign {
                ensure!(
                    current_flow_sign == flow_sign,
                    "PACs for process {} are a mix of inputs and outputs",
                    process_id
                );
            }
            flow_sign = Some(current_flow_sign);
```

Up to you which you think is clearer.